### PR TITLE
Capacity card for storage-dashboard

### PR DESCRIPTION
**Pull request by aruniiird**
This is basically a very minimal modification of wagner's PR#1723 to bring capacity card for storage-dashboard.

### CODE REVIEW

**Comment by aruniiird** [comment]
Screen shot:
persistent_storage

**Comment by aruniiird** [comment]
@cloudbehl , @rawagner , Can you guys please take a look at the changes and review the code.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  DashboardCardHeader,
+  DashboardCardTitle,
+} from '../../../../../../../public/components/dashboard/dashboard-card';
+import { CapacityBody, CapacityItem } from '../../../../../../../public/components/dashboard/capacity-card';
+import { withDashboardResources, DashboardItemProps } from '../../../../../../../public/components/dashboards-page/with-dashboard-resources';
+import { humanizeBinaryBytesWithoutB } from '../../../../../../../public/components/utils';
+import { getInstantVectorStats, getRangeVectorStats, GetStats } from '../../../../../../../public/components/graphs/utils';
+import { CAPACITY_CARD_QUERIES } from '../../../../constants/storage';
+
+const defaultQueries = {
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: @aruniiird please move these queries to `constant/storage.tsx`. How I did it in this PR.

https://github.com/openshift/console/pull/1803/files#diff-dd57dd229b99291aa21aa56a53d95b00

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  [CAPACITY_CARD_QUERIES.CAPACITY_TOTAL]: 'sum(node_filesystem_size_bytes)',
+  [CAPACITY_CARD_QUERIES.CAPACITY_USED]: '(sum(node_filesystem_size_bytes) - sum(node_filesystem_free_bytes))[60m:5m]',
+};
+
+const getLastStats = (response, getStats: GetStats): React.ReactText => {
+  const stats = getStats(response);
+  return stats.length > 0 ? stats[stats.length - 1].y : null;
+};
+
+const getQueries = () => {
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: We don't require this as we are not calling extensions in storage capacity.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  return _.defaults(pluginQueries, defaultQueries);
+};
+
+export const CapacityCard_: React.FC<DashboardItemProps> = ({
+  watchPrometheus,
+  stopWatchPrometheusQuery,
+  prometheusResults,
+}) => {
+  React.useEffect(() => {
+    const queries = getQueries();
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: We are not using any query from plugin.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    Object.keys(queries).forEach(key => watchPrometheus(queries[key]));
+    return () => Object.keys(queries).forEach(key => stopWatchPrometheusQuery(queries[key]));
+  }, [watchPrometheus, stopWatchPrometheusQuery]);
+
+  const queries = getQueries();
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: same as above

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  const storageUsed = prometheusResults.getIn([queries[CAPACITY_CARD_QUERIES.CAPACITY_USED], 'result']);
+  const storageTotal = prometheusResults.getIn([queries[CAPACITY_CARD_QUERIES.CAPACITY_TOTAL], 'result']);
+
+  return (
+    <DashboardCard>
+      <DashboardCardHeader>
+        <DashboardCardTitle>Cluster Capacity</DashboardCardTitle>
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]:
```suggestion
        <DashboardCardTitle>Capacity</DashboardCardTitle>
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            formatValue={humanizeBinaryBytesWithoutB}
+            isLoading={!(storageUsed && storageTotal)}
+          />
+        </CapacityBody>
+      </DashboardCardBody>
+    </DashboardCard>
+  );
+};
+
+export const CapacityCard = withDashboardResources(CapacityCard_);
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]:
```suggestion
export default withDashboardResources(CapacityCard);
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  plugins.registry.getDashboardsOverviewCapacityQueries().forEach(cq => {
+    const queryKey = cq.properties.queryKey;
+    if (!pluginQueries[queryKey]) {
+      pluginQueries[queryKey] = cq.properties.query;
+    }
+  });
+  return _.defaults(pluginQueries, defaultQueries);
+};
+
+export const CapacityCard_: React.FC<DashboardItemProps> = ({
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]:
```suggestion
export const CapacityCard: React.FC<DashboardItemProps> = ({
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+
+import * as plugins from '../../../../../../../public/plugins';
+import {
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]:
```suggestion
import { DashboardCard } from '@console/internal/components/dashboard/dashboard-card/card';
```

Comment (cloudbehl) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Use the same way to import all other internal imports.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+
+import * as plugins from '../../../../../../../public/plugins';
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Not required

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 10]
```diff
@@ -0,0 +1,73 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]:
```suggestion
import * as _ from 'lodash';
```

File: frontend/packages/ceph-storage-plugin/src/constants/storage.ts [hunk 11]
```diff
@@ -2,3 +2,8 @@ export const DATA_RESILIENCY_QUERIES = {
   CEPH_PG_CLEAN_AND_ACTIVE_QUERY: 'ceph_pg_clean and ceph_pg_active',
   CEPH_PG_TOTAL_QUERY: 'ceph_pg_total',
 };
+
+export const CAPACITY_CARD_QUERIES = {
+  CAPACITY_TOTAL: 'capacity_total',
+  CAPACITY_USED: 'capacity_used',
+}
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/constants/storage.ts]: missing new line

File: frontend/packages/ceph-storage-plugin/src/plugin.ts [hunk 12]
```diff
@@ -47,6 +47,17 @@ const plugin: Plugin<ConsumedExtensions> = [
...
   },
+  {
+    type: 'Dashboards/Card',
+    properties: {
+      tab: 'persistent-storage',
+      position: GridPosition.MAIN,
+      loader: () =>
+        import(
+          './components/dashboard-page/storage-dashboard/capacity-card/capacity-card' /* webpackChunkName: "ceph-storage-capacity-card" */
+        ).then((m) => m.CapacityCardWithResources),
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/plugin.ts]: This needs to be done once above comment is sorted
```suggestion
        ).then((m) => m.default),
```

**Comment by cloudbehl** [comment]
@aruniiird @rawagner Need to make changes to this PR once https://github.com/openshift/console/pull/1856 PR is merged.

**Comment by cloudbehl** [comment]
@aruniiird Build is failing can you fix the lint errors?
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_console/1854/pull-ci-openshift-console-master-frontend/5916

**Comment by aruniiird** [comment]
@cloudbehl , @rawagner , I have fixed the lint failures... but I'm not understanding the above failure. Any pointers?

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 1 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: As suggested in your PR, I have moved the variables into queries.ts, which helps to place labels and queries in a single file.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 2 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Ack, updated the PR

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 3 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Ack, updated the PR

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 4 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Done

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 5 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Acknowledged...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 6 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Acknowledged, made the changes in the latest PR...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 7 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Done...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 8 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Thanks Ankush, addressed in the latest PR...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 9 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Acknowledged...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 10 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Acknowledged...

File: frontend/packages/ceph-storage-plugin/src/constants/storage.ts (hunk 11 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/constants/storage.ts]: Corrected...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            formatValue={humanizeBinaryBytesWithoutB}
+            isLoading={!(storageUsed && storageTotal)}
+          />
+        </CapacityBody>
+      </DashboardCardBody>
+    </DashboardCard>
+  );
+};
+
+export const CapacityCardWithResources = withDashboardResources(CapacityCard);
```
[the remark this answers is not in the document]
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Done

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 12 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts]: Thanks for the sample code, was initially confused about how to do it... Changes made...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 14]
```diff
@@ -0,0 +1,67 @@
+import * as React from 'react';
+
+import {
+  DashboardCard,
+  DashboardCardBody,
+  DashboardCardHeader,
+  DashboardCardTitle,
+} from '@console/internal/components/dashboard/dashboard-card';
+
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: unnecessary empty line

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+import { CapacityBody, CapacityItem } from '@console/internal/components/dashboard/capacity-card';
+import {
+  withDashboardResources,
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: alpha

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/queries.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+export const enum CapacityQueryLabels {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/queries.ts]: should be moved here https://github.com/openshift/console/blob/master/frontend/packages/ceph-storage-plugin/src/constants/queries.ts

Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/queries.ts]: Acknowledged...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 15 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Sorry, I didn't get the comment... )=

Comment (rawagner) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: alphabetize the imports :)

Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Acknowledged

File: frontend/packages/ceph-storage-plugin/src/constants/queries.ts [hunk 17]
```diff
@@ -12,3 +12,14 @@ export const DATA_RESILIENCY_QUERIES = {
   [StorageDashboardQuery.CEPH_PG_CLEAN_AND_ACTIVE_QUERY]: 'ceph_pg_clean and ceph_pg_active',
   [StorageDashboardQuery.CEPH_PG_TOTAL_QUERY]: 'ceph_pg_total',
 };
+
+export const enum CapacityQueryLabels {
+  CAPACITY_TOTAL = 'capacity_total',
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/constants/queries.ts]: can you move these values to `StorageDashboardQuery` and follow the same pattern

Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/constants/queries.ts]: Oh! ok... Sure...

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx (hunk 14 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-card/capacity-card.tsx]: Acknowledged...

File: frontend/packages/ceph-storage-plugin/src/constants/queries.ts (hunk 17 above)
Comment (aruniiird) [reply frontend/packages/ceph-storage-plugin/src/constants/queries.ts]: Done...

**Comment by aruniiird** [comment]
Hey @rawagner , @cloudbehl ,
Please let me know if any further corrections have to be made...
Thanks

**Comment by aruniiird** [comment]
@cloudbehl , changes submitted... Please take a look...
Thanks

**Comment by rawagner** [comment]
@aruniiird please fix the linting errors

**Comment by aruniiird** [comment]
> @aruniiird please fix the linting errors

My bad, as it was a small change never expected a lint issue... Fixed... 
Thanks @rawagner

**Pull request merged**